Adds Brotli compression support for HTTP (via libbrotli)

Karm opened the pull request:
fixes #40692 
relates to #40533 

# Tested Platforms Checklist :heavy_check_mark: 
- [X] linux-x86_64
  - [X] MANDREL 23.1.2.0 JDK 21.0.2+13-LTS 
  - [X] MANDREL 23.0.4.1 JDK 17.0.11+9
- [X] linux-aarch64
  - [X] MANDREL 24.0.1.0 JDK 22.0.1+8
  - [X] MANDREL 23.1.3.1 JDK 21.0.3+9-LTS
  - [X] MANDREL 23.0.4.1 JDK 17.0.11+9
- [X] windows-x86_64
  - [X] MANDREL 23.1.3.1 JDK 21.0.3+9-LTS 
  - [X] MANDREL 23.0.4.1 JDK 17.0.11+9
- [X] osx-aarch64
  - [X] GRAALVM 24.0 JDK 22.0.1+8-jvmci-b01 

Ready for review.

Karm committed:
Adds Brotli support via native libbrotli

 * Integration tests for compressors
 * Adds Native image integration tests modules

cescoffier reviewed (commented):
LGTM!

Karm commented:
```
2024-05-22T11:42:29.0092701Z [INFO] asciidoctor: ERROR: telemetry-opentracing-to-otel-tutorial.adoc: line 306: dropping cells from incomplete row detected end of table
2024-05-22T11:42:29.0582466Z [INFO] Converted /home/runner/work/quarkus/quarkus/docs/target/asciidoc/sources/telemetry-opentracing-to-otel-tutorial.adoc
2024-05-22T11:42:29.0585271Z [ERROR] asciidoctor: ERROR: telemetry-opentracing-to-otel-tutorial.adoc: line 306: dropping cells from incomplete row detected end of table
```

Documentation error is unrelated to my PR. An error was merged with another PR.

Karm commented:
Documentation update preview: 

https://quarkus-pr-main-40750-preview.surge.sh/version/main/guides/http-reference#http-compression

Karm commented:
@zakkak Thx, commentary added [2ae2097](https://github.com/quarkusio/quarkus/pull/40750/commits/2ae2097d67ca304d4ed0bfbf16defb5275181047).

Karm commented:
@cescoffier This can be merged.

gsmet commented on .github/native-tests.json:
```diff
@@ -92,8 +92,8 @@
             "category": "HTTP",
-            "timeout": 110,
-            "test-modules": "elytron-resteasy, resteasy-jackson, elytron-resteasy-reactive, resteasy-mutiny, resteasy-reactive-kotlin/standard, vertx, vertx-http, vertx-web, vertx-web-jackson, vertx-graphql, virtual-http, rest-client, rest-client-reactive, rest-client-reactive-stork, rest-client-reactive-multipart, websockets, management-interface, management-interface-auth, mutiny-native-jctools ...
+            "timeout": 113,
```
I think you are a bit too optimistic about CI speed :). We usually add 5 minutes per added module.

gsmet commented on docs/src/main/asciidoc/http-reference.adoc:
```diff
@@ -69,9 +69,11 @@ The response body of a static resource is not compressed by default.
 If compression support is enabled then the response body is compressed if the `Content-Type` header derived from the file name of a resource is a compressed media type as configured via `quarkus.http.compress-media-types`.
 
-TIP: By default, the following list of media types is compressed: `text/html`, `text/plain`, `text/xml`, `text/css`, `text/javascript` and `application/javascript`.
+TIP: By default, the following list of media types is compressed: `text/html,text/plain,text/xml,text/css,text/javascript,application/javascript,application/graphql+json`. It means some other noteworthy MIME types such as `application/json,application/xhtml+xml` are NOT compressed by default.
```
I'm not convinced having a unique list with commas is more readable. I would prefer we keep the list as it was.

Should we consider adding the other two you mentioned to the default media types we compress? I'm not asking that for this PR but would it make sense to open an issue as a follow-up? (open question, I haven't looked at the details).

Also `s/MIME/media/` for consistency.

gsmet commented on extensions/vertx-http/runtime/src/main/java/io/quarkus/vertx/http/runtime/HttpBuildTimeConfig.java:
```diff
@@ -90,6 +90,22 @@ public class HttpBuildTimeConfig {
 
+    /**
+     * If user adds br, then brotli will be added to the list of supported compression algorithms.
+     * It implies loading libbrotli native library via JNI and in case of Native image,
```
```suggestion
     * It implies loading libbrotli native library via JNI and in case of native image,
```

gsmet commented on integration-tests/vertx-http-compressors/some/src/test/java/io/quarkus/compressors/it/RESTEndpointsTest.java:
```diff
@@ -0,0 +1,35 @@
+    @CsvSource(value = {
+    //@formatter:off
+    // Context | Accept-Encoding | Content-Encoding | Content-Length
+    "/yes/text | deflate,gzip,br | gzip             | 2414",
+    "/yes/text | deflate         | deflate          | 2402",
+    "/no/text  | deflate,gzip,br | null             | 6483",
+    "/yes/json | deflate         | deflate          | 2402",
+    "/no/json  | deflate,gzip,br | null             | 6483",
+    "/yes/xml  | deflate,gzip,br | gzip             | 2414",
+    "/no/xml   | deflate,gzip,br | null             | 6483",
```
Can't we test a marker in the binary rather than the content-length? This is going to be very brittle if some changes are made to any of the compression impl.

Karm replied:
The native part is built in the brotli4 repo, It's not imported. I had some `printf` in the C part there, so I know for sure it really loads what I think it does.

Karm replied:
@gsmet It kept nagging at me, so I conducted a quick experiment focused not on the compressor as such, but on the *default* parameters that are indeed changed in Netty.

```
Vert.x: 4.5.8, Brotli: 1.16.0, Netty: 4.1.110.Final
Uncompressed
content-length: 80948

content-encoding: gzip
content-length: 3530

content-encoding: deflate
content-length: 3518

content-encoding: br
content-length: 2758

Vert.x: 4.4.9, Brotli: 1.16.0, Netty: 4.1.108.Final
Uncompressed
content-length: 80948

content-encoding: gzip
content-length: 3530

content-encoding: deflate
content-length: 3518

content-encoding: br
content-length: 2758

Vert.x: 4.3.8, Brotli: 1.16.0, Netty: 4.1.87.Final
Uncompressed
content-length: 80948

content-encoding: gzip
content-length: 3530

content-encoding: deflate
content-length: 3518

content-encoding: br
content-length: 2757
```

Note that the  Netty: 4.1.87.Final  with Brotli compressed my test text to 2757 bytes whereas newer Netty to 2758.
I will add a tolerance window.

gsmet replied:
Again, can't we check a marker in the content instead? That's what I originally asked for and I don't want to be in the situation we already encountered with AWT and other things, where we have to update numeric values when a new implementation of Brotli appears.

Karm replied:
https://github.com/quarkusio/quarkus/pull/40750#issuecomment-2148046690

gsmet commented on extensions/resteasy-reactive/rest/deployment/src/main/java/io/quarkus/resteasy/reactive/server/deployment/CompressionScanner.java:
```diff
@@ -38,7 +38,8 @@ public List<HandlerChainCustomizer> scan(MethodInfo method, ClassInfo actualEndp
         }
 
-        AnnotationStore annotationStore = (AnnotationStore) methodContext.get(EndpointIndexer.METHOD_CONTEXT_ANNOTATION_STORE);
+        final AnnotationStore annotationStore = (AnnotationStore) methodContext
+                .get(EndpointIndexer.METHOD_CONTEXT_ANNOTATION_STORE);
```
I suppose you will have to push a new version to address some of my concerns. If so, please let's not include this change. If not, no need to bother.

gsmet reviewed (changes requested):
I added a couple of small comments.

Most are very minor but... I would really like to have the test fixed. I wouldn't want our tests to get broken every time GitHub Actions update their images and some compression impl changes.

Karm commented:
@gsmet  except for [compression numbers](https://github.com/quarkusio/quarkus/pull/40750#discussion_r1617417148), commit  191ba7c addresses your concerns.

galderz commented on integration-tests/vertx-http-compressors/app/src/test/java/io/quarkus/compressors/it/Testflow.java:
```diff
@@ -0,0 +1,112 @@
+                endpoint, acceptEncoding, contentEncoding, contentLength);
+        // RestAssured
+        // -----------
+        // Why not use RestAssured? Because it doesn't let you configure Accept-Encoding easily
```
Should an issue be raised with RestAssured to improve this?

Karm replied:
@galderz Being a test framework, I think it could do better at letting you tweak the headers manually. The topic is not new to the RestAssured GitHub though. I will make a TODO to follow up.

galderz commented on integration-tests/vertx-http-compressors/app/src/test/java/io/quarkus/compressors/it/Testflow.java:
```diff
@@ -0,0 +1,112 @@
+                // Why not use the client's built-in decompression support?
+                // Why you do decompression manually here?
+                // Because it then removes the original content-encoding header,
+                // and it fakes in a transfer-encoding header the server has never sent. Sad.
```
Should an issue be reported on Vert.x Web Client? /cc @vietj

vietj replied:
that might be done by Netty under the hood, what is the purpose of keeping the transfer-encoding untouched ? perhaps there is something wrong that should be fixed

Karm replied:
@galderz Thx for the review. I don't know. It could make sense for the general usage, but it makes it hard for testing a web server (Quarkus), where I explicitly want to manually fiddle with headers.

galderz reviewed (approved):
Looks good, just a couple of small comments to see if things can be improved upstream in Restassured and Vert.x Web Client. Great job @karm!

Karm commented:
@gsmet Ready to be merged.

Karm commented:
Oh my oh my.
> This branch has conflicts that must be resolved

This PR has been here for so long that apart from rebases there are conflicts now. 

As there is no new material feedback, I will merge it once I resolve the conflicts.

Karm commented:
Ad Data 7 module: #40960

Karm commented:
@gsmet 

> Again, can't we check a marker in the content instead? That's what I originally asked for

Checking the magic number at the beginning of the received file does not test that the compression took place as we expected, using the parameters we set. Note in the code that I am *not* comparing as `==`, test does `<=`, so nothing goes  south if the compression ratio gets magically better. If we suddenly botch passing parameters to Vert.x -> Netty, the compression would get worse result and fail.

I excruciatingly tested your hypothesis by testing numerous old versions across platforms. I did find that a very old Vert.x indeed showed 1 byte difference, better in fact, compression result. I accounted for this by adding a tolerance of 2 bytes.

I this PR, we have our webserver vendor's hats on and we are testing a web server. That's why I  manipulate headers in the test as that is what webserver test must do correctly. 
 
I don't want to be checking a magic number at the beginning of a binary and calling it a day, I want to see that the whole exchange makes sense and the compression works as expected. 

When I add dealing with chunked responses, this will get even more pertinent.
 
> and I don't want to be in the situation we already encountered with AWT and other things, where we have to update numeric values when a new implementation of Brotli appears.

Can you point to a single trouble with vastly more complex AWT test suite I hadn't fixed in a timely fashion? I maintain the tests I commit.

I had to refresh my memory, and I would like to add that the moment where pixel ranges suddenly didn't match was actually caused by the native-image using *a different library* for fonts processing, JDK in-tree built one versus the OS one. To learn about this change from the test suite is not without value. A change in JDK, given the AWT extension's nature, is again very valuable to learn about.

Bringing the argument home, if the compression ratio gets worse (e.g. by not passing correct compression level arguments), we want to know about it.

geoand commented:
Let's at least please up the commits and rebase onto the latest main

maxandersen commented on integration-tests/vertx-http-compressors/app/src/test/java/io/quarkus/compressors/it/Testflow.java:
```diff
@@ -0,0 +1,120 @@
+    // Depending on the defaults of the compression algorithm,
+    // the compressed content length may vary slightly between
+    // Vert.x/Netty versions over time.
+    public static final int COMPRESSION_TOLERANCE_BYTES = 2;
```
2 bytes is quite small for a 5x factor compression. would 2% tolerance be better / less likely to break ?

cescoffier replied:
Yes, having a reasonable tolerance window would mitigate the potential issue and still be accurate enough.

Karm replied:
Implemented in [cc361d1](https://github.com/quarkusio/quarkus/pull/40750/commits/cc361d12fcd99dd7c021a4ab92384c1f560e93e0).

gsmet commented:
@Karm FWIW, there was nothing personal in my comments.

I completely agree that you maintain the AWT tests and that you do it in a timely manner. But what if you are on PTO for 3 weeks or have more pressing things to do? We need to find a balance between being thorough and being brittle. And also how failures are actionable and by who. Because you, as any other developer, has the right to not be available to fix a test in the Quarkus test suite.

Now, if these actually allowed to catch important things to fix/adjust then it's all fine. My perception is that you had on a few occasions to spend some time maintaining some test values because of changes that were completely out of our control. Or because we wanted to test another platform/Java version/...

In this very case, we are testing a compression library that is out of our control. It will probably improve over time but really, there might be cases where it gets slightly worse for specific inputs for whatever reason (while improving overall).

If you want to make sure that a particular set of settings is taken into account then fine, but I'm not entirely sure testing the size is going to bring you that. Let's say they improve things well enough that even without your settings applied, you are below the threshold?

Anyway, what I wanted to emphasize is that it's not because it's fine now, that it will be fine along the life of Quarkus and the fact that Brotli compresses our text with 3 more bytes have very little value to us. And definitely not enough value for us to break our test suite.
FWIW, I didn't expect you to run tests on a gazillion of versions and sorry if you thought that's what I was asking. Because my problem is not Brotli on June 6th 2024, it's how to make sure Quarkus test suite remains sustainable in the long run if we follow this path for too many components.

I dismissed my review so feel free to merge this PR but I would still have a look at Max' comment, which results from a discussion we had just now.
It would make things less brittle while, I think, still testing what you want to test (or at least be an acceptable approximation).

Karm committed:
Adds 2% tolerance, i.e. compressor would have to get 2% or more WORSE at compressing our example to trigger a failure.

Karm commented:
> @Karm FWIW, there was nothing personal in my comments.
> 
> I completely agree that you maintain the AWT tests and that you do it in a timely manner. But what if you are on PTO for 3 weeks or have more pressing things to do? We need to find a balance between being thorough and being brittle. And also how failures are actionable and by who. Because you, as any other developer, have the right to not be available to fix a test in the Quarkus test suite.
> 
> Now, if these actually allowed to catch important things to fix/adjust then it's all fine. My perception is that you had on a few occasions to spend some time maintaining some test values because of changes that were completely out of our control. Or because we wanted to test another platform/Java version/...
> 
> In this very case, we are testing a compression library that is out of our control. It will probably improve over time but really, there might be cases where it gets slightly worse for specific inputs for whatever reason (while improving overall).
> 
> If you want to make sure that a particular set of settings is taken into account then fine, but I'm not entirely sure testing the size is going to bring you that. Let's say they improve things well enough that even without your settings applied, you are below the threshold?
> 
> Anyway, what I wanted to emphasize is that it's not because it's fine now, that it will be fine along the life of Quarkus and the fact that Brotli compresses our text with 3 more bytes have very little value to us. And definitely not enough value for us to break our test suite. FWIW, I didn't expect you to run tests on a gazillion of versions and sorry if you thought that's what I was asking. Because my problem is not Brotli on June 6th 2024, it's how to make sure Quarkus test suite remains sustainable in the long run if we follow this path for too many components.
> 
> I dismissed my review so feel free to merge this PR but I would still have a look at Max' comment, which results from a discussion we had just now. It would make things less brittle while, I think, still testing what you want to test (or at least be an acceptable approximation).

Thx for the additional explanation and especially for the time spent on the case. I appreciate it.

 > FWIW, I didn't expect you to run tests on a gazillion of versions and sorry if you thought that's what I was asking.
 
 That is on me. I was curious how stable those implementations behave. Whether they get any worse or better between versions historically.
 
 @gsmet @maxandersen @cescoffier 
 
 If for some reason the compressor gets wayyy *worse*, this is how the failure looks like:
 
```
 AssertionFailedError: Compression apparently failed: receivedLength: 2316 was
supposed to be less or equal to expectedLength: 1316 plus 2% tolerance, i.e. 1342. 
==> expected: <true> but was: <false>
```

maxandersen reviewed (approved):
looks good with the 2% threshold.
